fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,783 @@
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+const INTERVAL_COST_IN_BITS: usize = 64;
+
+/// Store blank size and position. Order by blank size.
+///
+/// A blank is an unoccupied space in the data.
+/// E.g. [100, 201] would have a `BlankAndPos{ pos: 0, blank_size: 100}`
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+struct BlankSizeAndPos {
+    blank_size: u128,
+    /// Position in the sorted data.
+    pos: usize,
+}
+impl BlankSizeAndPos {
+    fn new(blank: u128, pos: usize) -> Self {
+        BlankSizeAndPos {
+            blank_size: blank,
+            pos,
+        }
+    }
+}
+
+impl Ord for BlankSizeAndPos {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.blank_size.cmp(&other.blank_size)
+    }
+}
+impl PartialOrd for BlankSizeAndPos {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.blank_size.partial_cmp(&other.blank_size)
+    }
+}
+
+#[test]
+fn test_delta_and_pos_sort() {
+    let mut deltas: BinaryHeap<BlankSizeAndPos> = BinaryHeap::new();
+    deltas.push(BlankSizeAndPos::new(10, 1));
+    deltas.push(BlankSizeAndPos::new(100, 10));
+    deltas.push(BlankSizeAndPos::new(1, 10));
+    assert_eq!(deltas.pop().unwrap().blank_size, 100);
+    assert_eq!(deltas.pop().unwrap().blank_size, 10);
+}
+
+/// Put the deltas for the sorted values into a binary heap
+fn get_deltas(values_sorted: &[u128]) -> BinaryHeap<BlankSizeAndPos> {
+    let mut prev_opt = None;
+    let mut deltas: BinaryHeap<BlankSizeAndPos> = BinaryHeap::new();
+    for (pos, value) in values_sorted.iter().cloned().enumerate() {
+        let delta = if let Some(prev) = prev_opt {
+            value - prev
+        } else {
+            value + 1
+        };
+        // skip too small deltas
+        if delta > 2 {
+            deltas.push(BlankSizeAndPos::new(delta, pos));
+        }
+        prev_opt = Some(value);
+    }
+    deltas
+}
+
+struct BlankCollector {
+    blanks: Vec<BlankSizeAndPos>,
+    staged_blanks_sum: u128,
+}
+impl BlankCollector {
+    fn new() -> Self {
+        Self {
+            blanks: vec![],
+            staged_blanks_sum: 0,
+        }
+    }
+    fn stage_blank(&mut self, blank: BlankSizeAndPos) {
+        self.staged_blanks_sum += blank.blank_size - 1;
+        self.blanks.push(blank);
+    }
+    fn drain(&mut self) -> std::vec::Drain<'_, BlankSizeAndPos> {
+        self.staged_blanks_sum = 0;
+        self.blanks.drain(..)
+    }
+    fn staged_blanks_sum(&self) -> u128 {
+        self.staged_blanks_sum
+    }
+    fn num_blanks(&self) -> usize {
+        self.blanks.len()
+    }
+}
+
+/// Will collect blanks and add them to compact space if more bits are saved than cost from
+/// metadata.
+fn get_compact_space(values_sorted: &[u128], cost_per_blank: usize) -> CompactSpace {
+    let max_val = *values_sorted.last().unwrap_or(&0u128) + 1;
```

Comment:
that's not max_val. (tantivy uses this weird naming in max_doc to align itself with Lucene)